From 268e660aff62b7bb9e3897f4024ddea330a1967d Mon Sep 17 00:00:00 2001
From: Ze Tan <tanze@kylinos.cn>
Date: Tue, 4 Aug 2026 16:57:12 +0800
Subject: [PATCH 4/5] ksmbd: fail initialization when procfs setup fails

ksmbd_proc_init() cleans up partial state on failure but does not report
the error. Module initialization then continues and later statistics
updates access percpu counters that were never initialized or were
already destroyed.

Return procfs setup errors to ksmbd_server_init(), also check creation of
the global sessions entry, and abort module initialization after cleaning
up partial procfs state.

Signed-off-by: Ze Tan <tanze@kylinos.cn>
---
 fs/smb/server/misc.h   |  4 ++--
 fs/smb/server/proc.c   | 18 +++++++++++-------
 fs/smb/server/server.c | 11 +++++++++--
 3 files changed, 22 insertions(+), 11 deletions(-)

diff --git a/fs/smb/server/misc.h b/fs/smb/server/misc.h
index 3909104e18ad..c7b063f571a7 100644
--- a/fs/smb/server/misc.h
+++ b/fs/smb/server/misc.h
@@ -43,7 +43,7 @@ struct ksmbd_const_name {
 	const char *name;
 };
 
-void ksmbd_proc_init(void);
+int ksmbd_proc_init(void);
 void ksmbd_proc_cleanup(void);
 void ksmbd_proc_reset(void);
 struct proc_dir_entry *ksmbd_proc_create(const char *name,
@@ -59,7 +59,7 @@ void ksmbd_proc_show_const_name(struct seq_file *m,
 				int count,
 				unsigned int const_value);
 #else
-static inline void ksmbd_proc_init(void) {}
+static inline int ksmbd_proc_init(void) { return 0; }
 static inline void ksmbd_proc_cleanup(void) {}
 static inline void ksmbd_proc_reset(void) {}
 #endif
diff --git a/fs/smb/server/proc.c b/fs/smb/server/proc.c
index 101a2cc45a44..adcd397cdb77 100644
--- a/fs/smb/server/proc.c
+++ b/fs/smb/server/proc.c
@@ -106,29 +106,33 @@ void ksmbd_proc_reset(void)
 		percpu_counter_set(&ksmbd_counters.counters[i], 0);
 }
 
-void ksmbd_proc_init(void)
+int ksmbd_proc_init(void)
 {
 	int i;
-	int retval;
+	int ret = -ENOMEM;
 
 	ksmbd_proc_fs = proc_mkdir("fs/ksmbd", NULL);
 	if (!ksmbd_proc_fs)
-		return;
+		return -ENOMEM;
 
 	if (!proc_mkdir_mode("sessions", 0400, ksmbd_proc_fs))
 		goto err_out;
 
 	for (i = 0; i < ARRAY_SIZE(ksmbd_counters.counters); i++) {
-		retval = percpu_counter_init(&ksmbd_counters.counters[i], 0, GFP_KERNEL);
-		if (retval)
+		ret = percpu_counter_init(&ksmbd_counters.counters[i], 0,
+					  GFP_KERNEL);
+		if (ret)
 			goto err_out;
 	}
 
-	if (!ksmbd_proc_create("server", proc_show_ksmbd_stats, NULL))
+	if (!ksmbd_proc_create("server", proc_show_ksmbd_stats, NULL)) {
+		ret = -ENOMEM;
 		goto err_out;
+	}
 
 	ksmbd_proc_reset();
-	return;
+	return 0;
 err_out:
 	ksmbd_proc_cleanup();
+	return ret;
 }
diff --git a/fs/smb/server/server.c b/fs/smb/server/server.c
index 960c4c897c11..8d8e857b1d57 100644
--- a/fs/smb/server/server.c
+++ b/fs/smb/server/server.c
@@ -603,8 +603,15 @@ static int __init ksmbd_server_init(void)
 		return ret;
 	}
 
-	ksmbd_proc_init();
-	create_proc_sessions();
+	ret = ksmbd_proc_init();
+	if (ret)
+		goto err_unregister;
+
+	ret = create_proc_sessions();
+	if (ret) {
+		ksmbd_proc_cleanup();
+		goto err_unregister;
+	}
 
 	ksmbd_server_tcp_callbacks_init();
 
-- 
2.43.0

